domains/client/block-preprocessor/src/runtime_api_light.rs

```diff
@@ -0,0 +1,170 @@
+}
+
+impl<Block, Executor> Core<Block> for RuntimeApiLight<Executor>
+where
+    Block: BlockT,
+    Executor: CodeExecutor + RuntimeVersionOf,
+{
+    fn __runtime_api_internal_call_api_at(
+        &self,
+        _at: <Block as BlockT>::Hash,
+        _context: ExecutionContext,
+        params: Vec<u8>,
+        fn_name: &dyn Fn(RuntimeVersion) -> &'static str,
+    ) -> Result<Vec<u8>, ApiError> {
+        self.dispatch_call(fn_name, params)
+    }
+}
+
+impl<Block, Executor> MessengerApi<Block, NumberFor<Block>> for RuntimeApiLight<Executor>
+where
+    Block: BlockT,
+    NumberFor<Block>: Codec,
+    Executor: CodeExecutor + RuntimeVersionOf,
+{
+    fn __runtime_api_internal_call_api_at(
+        &self,
+        _at: <Block as BlockT>::Hash,
+        _context: ExecutionContext,
+        params: Vec<u8>,
+        fn_name: &dyn Fn(RuntimeVersion) -> &'static str,
+    ) -> Result<Vec<u8>, ApiError> {
+        self.dispatch_call(fn_name, params)
+    }
+}
+
+impl<Executor> FetchRuntimeCode for RuntimeApiLight<Executor> {
+    fn fetch_runtime_code(&self) -> Option<Cow<'static, [u8]>> {
+        Some(self.runtime_code.clone())
+    }
+}
+
+impl<Executor> RuntimeApiLight<Executor>
+where
+    Executor: CodeExecutor + RuntimeVersionOf,
+{
+    #[allow(unused)]
+    pub fn new(executor: Arc<Executor>, runtime_code: Cow<'static, [u8]>) -> Self {
+        Self {
+            executor,
+            runtime_code,
+        }
+    }
+
+    fn runtime_code(&self) -> RuntimeCode<'_> {
+        let code_hash = sp_core::Blake2Hasher::hash(&self.runtime_code);
+        RuntimeCode {
+            code_fetcher: self,
+            heap_pages: None,
+            hash: code_hash.encode(),
+        }
+    }
+
+    fn runtime_version(&self) -> Result<RuntimeVersion, ApiError> {
+        let mut ext = BasicExternalities::new_empty();
+        self.executor
+            .runtime_version(&mut ext, &self.runtime_code())
+            .map_err(|err| ApiError::Application(Box::new(err)))
+    }
+
+    fn dispatch_call(
+        &self,
+        fn_name: &dyn Fn(RuntimeVersion) -> &'static str,
+        input: Vec<u8>,
+    ) -> Result<Vec<u8>, ApiError> {
+        let runtime_version = self.runtime_version()?;
+        let fn_name = fn_name(runtime_version);
+        let mut ext = BasicExternalities::new_empty();
+        self.executor
+            .call(
+                &mut ext,
+                &self.runtime_code(),
+                fn_name,
+                &input,
+                false,
+                CallContext::Offchain,
+            )
+            .0
+            .map_err(|err| {
+                ApiError::Application(format!("Failed to invoke call to {fn_name}: {err}").into())
+            })
+    }
+}
+
+impl<Block, Executor> StateRootExtractor<Block> for RuntimeApiLight<Executor>
+where
+    Block: BlockT,
+    Executor: CodeExecutor + RuntimeVersionOf,
+{
+    fn extract_state_roots(
+        &self,
+        block_hash: Block::Hash,
```

Comment:
So all the block_hash is unused for the stateless impl, can we make it more explicit so the user of this API doesn't have to worry about it at all? For example, also add the stateless version into the trait something like:

```Rust
/// Trait to extract XDM state roots from the Extrinsic.
pub trait StateRootExtractor<Block: BlockT> {
    /// Extracts the state roots from the extrinsic using the client.
    fn extract_state_roots(
        &self,
        at: Block::Hash,
        ext: &Block::Extrinsic,
    ) -> Result<ExtractedStateRoots<Block>, ApiError>;

    /// Extracts the state roots from the extrinsic using the stateless runtime.
    fn extract_state_roots_stateless(
        &self,
        ext: &Block::Extrinsic,
    ) -> Result<ExtractedStateRoots<Block>, ApiError>;
}
```
Don't have to be done in this PR, if you agree, I can work on top of it later as I find there are left-overs anyway.

Reply:
Hmm, this would require the stateless version of `preprocess_primary_block(primary_hash)`, nvm, I can just pass a dumb `domain_hash`.

Reply:
hmm why, It simpler to have one API. Stateless impl will ignore it but api caller will still value it. I would leave this as is.

Reply:
Ideally, if an argument is unused, then it shouldn't be required at all from API user's perpective. It's fine for now as I don't see an easy and nice way to fix it.

Reply:
why do you want to seperate the call into stateful and stateless? The api should not leak such info anyway. Then `at` is used for when required.

Reply:
Because it's not necessary when calling it from the verifier side. If unused, then it shouldn't be required ideally in general. Furthermore, now we only include the domain parent hash for InitializeBlock, it's not available for ApplyExtrinsic and FinalizeBlock, which means some dumb value will be used when it's unavailable if this argument is always required.

Reply:
Why is not necessary ?
On verifier side, this should be domain parent hash no ?

Imo, this is impleaation specific detail for single trait function. I'm sure I like the proposed trait def tbh. That defeats the purpose of abstracting it out.

Having said that, if you think the best approach is what you suggested, I'm not going to block it ,👍